orttraining/orttraining/test/gradient/gradient_ops_test.cc

```diff
@@ -1492,8 +1492,12 @@ void TestDropoutGradOp(float ratio, TensorShape& x_shape, bool default_ratio = t
                                                   true, false, true, false});
   if (!default_ratio) {
     test.AddInput<float>("ratio", {1}, ratio_data);
+  } else {
+    test.AddMissingOptionalInput<bool>();
   }
 
+  test.AddInput("training_mode", {}, {true});
```

Comment:
>training_mode [](start = 17, length = 13)

the DropoutGrad kernel also need to handle the "false" training mode, for the sake of completeness. 

dX will be equal to dY in such case.

Reply:
Please also a few test cases to cover this scenario.

---
In reply to: [431343860](https://github.com/microsoft/onnxruntime/pull/4052#discussion_r431343860) [](ancestors = 431343860)

Reply:
@SherlockNoMad I left out training_mode from the grad kernel on purpose because when grad kernel is invoked we know it will be when training_mode is true. Is there a case where we do training with training_mode = false?

Reply:
I also thought that there wasn't a scenario where a Grad kernel is called during inference. In that case, I'd argue that if `DropoutGrad` is called with `training_mode=false` then we should throw an exception as opposed to return the input as output

Reply:
I shared the same thought with you at the very begining, but I realized that there might be some niche scenario that use might want to freeze some layers during the training (I may be making this up, but having training_mode as an input of Dropout Spec allows such case to happen). 

So we might just as well support this "training_mode=false" for the DropoutGrad. It's not too complicated to implement and doesn't cost the perf.